yapapi/strategy.py

```diff
@@ -21,10 +23,130 @@
+    PROP_PAYMENT_TIMEOUT_SEC: PropValueRange(DEFAULT_PAYMENT_TIMEOUT_SEC, None),
+    PROP_DEBIT_NOTE_ACCEPTANCE_TIMEOUT: PropValueRange(
+        DEFAULT_DEBIT_NOTE_ACCEPTANCE_TIMEOUT_SEC, None
+    ),
+}
+
+logger = logging.getLogger(__name__)
+
 
 class MarketStrategy(DemandDecorator, abc.ABC):
     """Abstract market strategy."""
 
+    acceptable_prop_value_ranges: Dict[str, PropValueRange]
+
+    def set_prop_value_ranges_defaults(
+        self, acceptable_prop_value_ranges: Dict[str, PropValueRange]
+    ) -> None:
+        try:
+            value_ranges = self.acceptable_prop_value_ranges
+            for key, value in acceptable_prop_value_ranges.items():
+                if key not in value_ranges:
+                    value_ranges[key] = value
+        except AttributeError:
+            self.acceptable_prop_value_ranges = acceptable_prop_value_ranges.copy()
+
+    async def respond_to_provider_offer(
+        self,
+        our_demand: DemandBuilder,
+        provider_offer: rest.market.OfferProposal,
+    ) -> DemandBuilder:
+        # Create a new DemandBuilder with a response to a provider offer.
+        updated_demand = deepcopy(our_demand)
+
+        # Set default property value ranges if they were not set in the market strategy.
+        self.set_prop_value_ranges_defaults(DEFAULT_PROPERTY_VALUE_RANGES)
+
+        # only enable mid-agreement-payments when we need a longer expiration
+        # and when the provider supports it
+        activity = Activity.from_properties(our_demand.properties)
+
+        assert activity.expiration
+        expiration_secs = round((activity.expiration - datetime.now(timezone.utc)).total_seconds())
+        trigger_mid_agreement_payments = expiration_secs >= float(
+            MIN_EXPIRATION_FOR_MID_AGREEMENT_PAYMENTS
+        )
+
+        mid_agreement_payments_enabled = (
+            PROP_DEBIT_NOTE_INTERVAL_SEC in provider_offer.props and trigger_mid_agreement_payments
+        )
+
+        if mid_agreement_payments_enabled:
+            logger.debug(
+                "Enabling mid-agreement payments mechanism "
+                "as the expiration set to %ss (more than %ss).",
+                expiration_secs,
+                MIN_EXPIRATION_FOR_MID_AGREEMENT_PAYMENTS,
+            )
+        elif trigger_mid_agreement_payments:
+            logger.debug(
+                "Expiration of %ss (more than our minimum for MAP: %ss) while negotiating "
+                "with a provider unaware of mid-agreeement-payments.",
+                expiration_secs,
+                MIN_EXPIRATION_FOR_MID_AGREEMENT_PAYMENTS,
+            )
+
+        # Update response by either accepting proposed values or by proposing ours
+        # only set mid-agreement payments values if we're agreeing to them
+        for prop_key, acceptable_range in self.acceptable_prop_value_ranges.items():
+            prop_value = provider_offer.props.get(prop_key)
+            if prop_value and (
```

Comment:
```suggestion
            if prop_value is not None and (
```
I'm not sure about this, but I think it's safer and more readable? Although I might be missing something.

Reply:
definitely a good point, I'm including it into the commit